src/memory.rs

```diff
@@ -59,6 +59,8 @@ pub struct Memory<'tcx> {
     /// Function "allocations". They exist solely so pointers have something to point to, and
     /// we can figure out what they point to.
     functions: HashMap<AllocId, FunctionDefinition<'tcx>>,
+    /// Inverse map of `functions` so we don't allocate a new pointer every time we need one
+    function_definitions: HashMap<FunctionDefinition<'tcx>, AllocId>,
```

Comment:
I'd prefer a name like `function_allocs` or `function_alloc_cache` since the values we get out of it are allocations.